### PR TITLE
wip : add WaVLM-Base model to PyanNet.py in replacement of SincNet

### PR DESCRIPTION
**Pull request by SevKod**
Added WavLM-Base model which replaces the SincNet feature extraction model within the PyanNet architecture (loaded outside of the class from HuggingFace.co).

### CODE REVIEW

File: pyannote/audio/models/segmentation/PyanHugg.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+        super().__init__(sample_rate=sample_rate, num_channels=num_channels, task=task)
+        
+        selfsupervised = merge_dict(self.SSL_DEFAULTS, selfsupervised)
+        sincnet = merge_dict(self.SINCNET_DEFAULTS, sincnet)
+        sincnet["sample_rate"] = sample_rate
+        lstm = merge_dict(self.LSTM_DEFAULTS, lstm)
+        lstm["batch_first"] = True
+        linear = merge_dict(self.LINEAR_DEFAULTS, linear)
+        if (selfsupervised["model"] == "sincnet") :
```
Comment (hbredin) [inline pyannote/audio/models/segmentation/PyanHugg.py:93]: I would remove support for `SincNet` completely to avoid any confusion.

File: pyannote/audio/models/blocks/selfsup.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        self.config = config #Assign the configuration
+        SelfSupModel.__name__ = self.model_name #Assign name of the class
+        
+        if name is "WAVLM_BASE" or name is "WAVLM_LARGE": #Only wavlm_model has two additional arguments
+            model = wavlm_model(**config)
+        else:
+            model = wav2vec2_model(**config)
+        model.load_state_dict(ordered_dict) #Assign state dict to the model
+        
+        if finetune:
```
Comment (hbredin) [inline pyannote/audio/models/blocks/selfsup.py:110]: Peux-tu m'expliquer pourquoi cela est nécessaire ?

Comment (hbredin) [reply pyannote/audio/models/blocks/selfsup.py:110]: Je parle du passage au mode `eval` dans le cas où le WavLM est gelé.

Comment (SevKod) [reply pyannote/audio/models/blocks/selfsup.py:110]: J'avais cru comprendre que le .eval() était pertinent lorsque certains modules tel que des couches de Dropout, sont présentes dans le modèle en question que l'on souhaite passer en mode inférence (le cas pour WavLM). J'avais lu ce post qui conseillait l'utilisation des deux :

 https://stackoverflow.com/questions/55627780/evaluating-pytorch-models-with-torch-no-grad-vs-model-eval

Après, de souvenirs, je n'avais pas identifié de quelconque changement au niveau des features entre le ".eval()" et le "no_grad". Apparemment, le .eval() consommerait plus de mémoire que le no_grad aussi... Donc, pourquoi pas l'enlever. C'était aussi pour étudier les changements entre le .eval() et le .train() quand je voulais voir ce qui se passait au niveau du finetuning.

Comment (hbredin) [reply pyannote/audio/models/blocks/selfsup.py:110]: `model.eval()` et `torch.no_grad()` ont deux rôles bien différents.

* `torch.no_grad()` supprimer le calcul du gradient des couches concernées et est donc utile quand tu veux geler une partie du réseau.
* `model.eval()` passe les couches qui ont un comportement particulier lors de l'apprentissage (e.g. `dropout` qui désactive aléatoirement certains poids, ou `batchnorm` qui calcule une moyenne des données qui la traverse) en mode _inférence_ pour supprimer tout cet aléa.

En résumé, il ne faut pas utiliser ni `model.eval()` ni `model.train()` pour contrôler si tu finetunes ou non la partie _feature extraction_. Il faut seulement utiliser `torch.no_grad()` (pour geler) ou pas (pour finetuner).

Le passage en mode `eval` ou `train` est effectué automatiquement par `pytorch-lightning` lors des phases de validation et d'apprentissage.

File: pyannote/audio/models/blocks/selfsup.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            self.W = nn.Parameter(torch.randn(config['encoder_num_layers'])) #Set max number of learnable weights
+            
+            self.average_layers = list(range(config['encoder_num_layers']))
+            
+            self.layer = config['encoder_num_layers']
+            
+        if finetune: #Finetuning not working
+            print("Self-supervised model is unfrozen.")      
+            #config['encoder_ff_interm_dropout'] = 0.3
+            config['encoder_layer_norm_first'] = True
```
Comment (hbredin) [inline pyannote/audio/models/blocks/selfsup.py:91]: Can you explain?

Comment (SevKod) [reply pyannote/audio/models/blocks/selfsup.py:91]: The issue regarding the funetuning of WavLM seemed similar to a normalization issue that occured during the feature extraction process. If gradient is computed during training, validation will extract feature vectors that are almost identical amongst each frames of the input audio. I assume that it might be the reason why validation does not seem to improve (or change) during training (but I might be completely wrong on this...). Since this problem seemed similar to the one I encountered with WavLM from back a few months (and has been fixed), where features were the also the same amongst the frames, I tried applying a normalization step to see if the behavior of the features extracted would change. Did not seem to be the case... It is one of the many things I tried but forgot to remove when pushing codes ^^
